File: android/src/main/java/com/verifai/reactnative/liveness/LivenessModule.java
package com.verifai.reactnative.liveness;

import android.app.Activity;
import android.os.Build;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.module.annotations.ReactModule;
import com.google.gson.Gson;
import com.verifai.liveness.VerifaiLiveness;
import com.verifai.liveness.VerifaiLivenessCheckListener;
import com.verifai.liveness.checks.CloseEyes;
import com.verifai.liveness.checks.FaceMatching;
import com.verifai.liveness.checks.Speech;
import com.verifai.liveness.checks.Tilt;
import com.verifai.liveness.checks.VerifaiLivenessCheck;
import com.verifai.liveness.result.VerifaiLivenessCheckResults;
import com.verifai.reactnative.ConvertDataUtilities;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

/**
 * Wrapper class for the liveness module
 * Callbacks have to be set one by one because:
 * https://reactnative.dev/docs/native-modules-android#callbacks
 * Maybe this can be done differently when TurboModules are out
 */
@ReactModule(name = LivenessModule.NAME)
public class LivenessModule extends ReactContextBaseJavaModule {
    public static final String NAME = "Liveness";
    public static final String TAG = "V-LIVE";

    public LivenessModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    private Callback _onSuccess = null;

    @ReactMethod
    public void setOnSuccess(Callback onSuccess) {
        _onSuccess = onSuccess;
    }

    private Callback _onError = null;

    @ReactMethod
    public void setOnError(Callback onError) {
        _onError = onError;
    }

    private final ConvertDataUtilities utils = new ConvertDataUtilities();

    @Override
    @NonNull
    public String getName() {
        return NAME;
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    @ReactMethod
    public void start() {
        Activity activity = getCurrentActivity();
        if (activity == null) {
            Log.e(TAG, "No activity running");
            return;
        }
        if (_onSuccess == null) {
            Log.e(TAG, "No onSuccess callback has been set");
            return;
        }
        if (_onError == null) {
            Log.e(TAG, "No onError callback has been set");
            return;
        }

        VerifaiLivenessCheckListener resultListener = new VerifaiLivenessCheckListener() {
            @Override
            public void onResult(@NonNull VerifaiLivenessCheckResults results) {
                try {
                    Gson gson = new Gson();
                    String jsonResult = gson.toJson(results);
                    WritableMap returnMap = utils.convertJsonToMap(new JSONObject(jsonResult));
                    _onSuccess.invoke(returnMap);
                } catch (JSONException e) {
                    onError(new Exception("Data conversion failed"));
                }
            }

            @Override
            public void onError(Throwable throwable) {
                _onError.invoke(throwable.getMessage());
            }
        };
        ArrayList<VerifaiLivenessCheck> checks = new ArrayList<>();
//        if (result != null) {
//            checks.add(new FaceMatching(this, Objects.requireNonNull(result.getFrontImage())));
//        }
        checks.add(new Tilt(activity, -25));
        checks.add(new CloseEyes(activity, 2));
        VerifaiLiveness.start(activity, checks, resultListener);
    }
}

File: android/src/main/java/com/verifai/reactnative/LivenessModule.kt
package com.verifai.reactnative

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.module.annotations.ReactModule
import com.verifai.liveness.pub.LivenessCheckListener
import com.verifai.liveness.pub.LivenessConfiguration
import com.verifai.liveness.pub.VerifaiLiveness
import com.verifai.liveness.pub.checks.LivenessCheck
import com.verifai.liveness.pub.result.LivenessCheckResults
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.jsonObject

/**
 * Wrapper class for the liveness module
 */
@ReactModule(name = LivenessModule.NAME)
class LivenessModule(reactContext: ReactApplicationContext?) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String {
        return NAME
    }

    private val json = Json {
        encodeDefaults = true
    }

    @ReactMethod
    fun configure(args: ReadableMap, promise: Promise) {
        try {
            val configuration: LivenessConfiguration = json.decodeFromJsonElement(args.toJsonObject())
            VerifaiLiveness.configure(configuration)
            promise.resolve(null)
        } catch (e: Throwable) {
            promise.reject(e)
        }
    }

    @ReactMethod
    fun start(args: ReadableArray, promise: Promise) {
        try {
            val livenessChecks: List<LivenessCheck> = json.decodeFromJsonElement(args.toJsonArray(true))
            val resultListener = object : LivenessCheckListener {
                override fun onResult(results: LivenessCheckResults) {
                    try {
                        promise.resolve(json.encodeToJsonElement(results).jsonObject.toRNMap())
                    } catch (e: Throwable) {
                        promise.reject(Error.ResultConversion, e)
                    }
                }

                override fun onError(e: Throwable) {
                    promise.reject(Error.Liveness, e)
                }
            }
            VerifaiLiveness.start(currentActivity!!, resultListener, livenessChecks)
        } catch (e: Throwable) {
            promise.reject(Error.Unhandled, e)
        }
    }

    companion object {
        const val NAME = "Liveness"
    }
}
